src/core/message-parser.ts

```diff
@@ -41,17 +62,71 @@ export function parseMessage(roomId: string | undefined, message: Ably.InboundMe
     throw new Ably.ErrorInfo(`received incoming message without extras`, 50000, 500);
   }
 
-  if (!messageCreatedMessage.extras.timeserial) {
+  if (!messageCreatedMessage.serial) {
     throw new Ably.ErrorInfo(`received incoming message without timeserial`, 50000, 500);
   }
 
-  return new DefaultMessage(
-    messageCreatedMessage.extras.timeserial,
-    messageCreatedMessage.clientId,
+  let operationDetails: MessageDetails | undefined;
+  if (messageCreatedMessage.operation) {
+    operationDetails = {
+      description: messageCreatedMessage.operation.description,
+      metadata: messageCreatedMessage.operation.metadata,
+    };
+  }
+
+  const newMessage: ChatMessageFields = {
+    timeserial: messageCreatedMessage.serial,
+    clientId: messageCreatedMessage.clientId,
     roomId,
-    messageCreatedMessage.data.text,
-    new Date(messageCreatedMessage.timestamp),
-    messageCreatedMessage.data.metadata ?? {},
-    messageCreatedMessage.extras.headers ?? {},
+    text: messageCreatedMessage.data.text,
+    createdAt: new Date(messageCreatedMessage.timestamp),
+    metadata: messageCreatedMessage.data.metadata ?? {},
+    headers: messageCreatedMessage.extras.headers ?? {},
+  };
+
+  switch (messageCreatedMessage.action) {
+    case 'MESSAGE_CREATE': {
```

Comment:
Should this use the [Ably.MessageAction[s]](https://github.com/ably/ably-js/pull/1888/files/2a59fb45d8243f022677143af551217e71eefaf9#r1796692076) thing?

Reply:
I'm not sure what you mean, actions is defined above as `Ably.MessageAction` already?

Reply:
I meant should it be `Ably.MessageActions.MESSAGE_CREATE` instead of directly a string?

Reply:
This doesn't work owing to the fact that we haven't exposed this as an enum. But, I have created an enum type we can use, just like we do with Presence :)